File: solutions/heap/295-find-median-from-data-stream.py
"""
# Difficulty: Hard

# 295. Find Median from Data Stream

The median is the middle value in an ordered integer list. If the size of the list is even, there is no middle value, and the median is the mean of the two middle values.

Implement the MedianFinder class:
- MedianFinder() initializes the MedianFinder object.
- void addNum(int num) adds the integer num from the data stream to the data structure.
- double findMedian() returns the median of all elements so far.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>addNum(1):</dd>
<dt>Output:</dt>
<dd>left=[1], right=[]</dd>
<dt>Explanation:</dt>
<dd>Median after adding [1,2] is 1.5, after [1,2,3] is 2</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
Use two heaps to divide numbers into two halves: max-heap for smaller half, min-heap for larger half. The median is always at the top of one or both heaps.

### APPROACH:
1. **Two heaps**: max-heap (left half), min-heap (right half)
2. **Balance heaps**: Keep sizes equal or left heap has 1 more
3. **Add number**:
   - Add to left heap, then move largest to right heap
   - If right > left, move smallest from right to left
4. **Find median**:
   - If sizes equal: average of both tops
   - Else: top of left heap

### WHY THIS WORKS:
- Max-heap stores smaller half (can get largest quickly)
- Min-heap stores larger half (can get smallest quickly)
- Tops of heaps are always the middle elements
- Balancing ensures O(1) median access

### EXAMPLE WALKTHROUGH:
```
addNum(1):
  left=[1], right=[]
  median = 1

addNum(2):
  left=[1], right=[2]
  median = (1+2)/2 = 1.5

addNum(3):
  left=[2,1], right=[3]
  median = 2 (top of left heap)

addNum(4):
  left=[2,1], right=[3,4]
  median = (2+3)/2 = 2.5
```

### TIME COMPLEXITY:
- addNum(): O(log n) - heap operations
- findMedian(): O(1) - just access heap tops

### SPACE COMPLEXITY:
O(n)
Store all n numbers across two heaps

### EDGE CASES:
- Single element
- Two elements
- Negative numbers
- Duplicate values
- Large data stream

</details>
"""

File: solutions/heap/295-find-median-from-data-stream.js
I'll help you create a complete JavaScript implementation of a MedianFinder class that efficiently finds the median of a data stream. Here's the solution:

```javascript
/**
 * MedianFinder class for finding the median in a data stream
 * Uses two heaps (max heap for lower half, min heap for upper half)
 * to maintain the median efficiently
 */
class MedianFinder {
    constructor() {
        // Initialize two heaps
        this.maxHeap = []; // For lower half of numbers
        this.minHeap = []; // For upper half of numbers
    }

    /**
     * Adds a number to the data structure
     * @param {number} num - The number to add
     */
    addNum(num) {
        // If maxHeap is empty or num is less than maxHeap's top, add to maxHeap
        if (this.maxHeap.length === 0 || num < -this.maxHeap[0]) {
            this.maxHeap.push(-num);
            this.maxHeapify(this.maxHeap.length - 1);
        } else {
            this.minHeap.push(num);
            this.minHeapify(this.minHeap.length - 1);
        }

        // Balance heaps if necessary
        this.balanceHeaps();
    }

    /**
     * Returns the median of all numbers seen so far
     * @return {number} The median value
     */
    findMedian() {
        if (this.maxHeap.length === this.minHeap.length) {
            return ((-this.maxHeap[0] + this.minHeap[0]) / 2);
        } else {
            return -this.maxHeap[0];
        }
    }

    /**
     * Maintains heap property for max heap
     * @param {number} index - Index of the element to heapify
     */
    maxHeapify(index) {
        while (index > 0) {
            const parent = Math.floor((index - 1) / 2);
            if (-this.maxHeap[parent] < -this.maxHeap[index]) {
                [this.maxHeap[parent], this.maxHeap[index]] = 
                    [this.maxHeap[index], this.maxHeap[parent]];
                index = parent;
            } else {
                break;
            }
        }
    }

    /**
     * Maintains heap property for min heap
     * @param {number} index - Index of the element to heapify
     */
    minHeapify(index) {
        while (index > 0) {
            const parent = Math.floor((index - 1) / 2);
            if (this.minHeap[parent] > this.minHeap[index]) {
                [this.minHeap[parent], this.minHeap[index]] = 
                    [this.minHeap[index], this.minHeap[parent]];
                index = parent;
            } else {
                break;
            }
        }
    }

    /**
     * Balances the two heaps to maintain the median property
     * Ensures maxHeap has either equal or one more element than minHeap
     */
    balanceHeaps() {
        // If the difference in sizes is more than 1, rebalance
        while (this.maxHeap.length > this.minHeap.length + 1) {
            this.minHeap.push(-this.maxHeap[0]);
            this.maxHeap[0] = -this.maxHeap[this.maxHeap.length - 1];
            this.maxHeap.pop();
            this.siftDownMax(0);
            this.minHeapify(this.minHeap.length - 1);
        }

        while (this.minHeap.length > this.maxHeap.length) {
            this.maxHeap.push(-this.minHeap[0]);
            this.minHeap[0] = this.minHeap[this.minHeap.length - 1];
            this.minHeap.pop();
            this.siftDownMin(0);
            this.maxHeapify(this.maxHeap.length - 1);
        }
    }

    /**
     * Sifts down an element in the max heap
     * @param {number} index - Starting index for sift down operation
     */
    siftDownMax(index) {
        const length = this.maxHeap.length;
        while (true) {
            let largest = index;
            const left = 2 * index + 1;
            const right = 2 * index + 2;

            if (left < length && -this.maxHeap[left] > -this.maxHeap[largest]) {
                largest = left;
            }
            if (right < length && -this.maxHeap[right] > -this.maxHeap[largest]) {
                largest = right;
            }

            if (largest === index) break;

            [this.maxHeap[index], this.maxHeap[largest]] = 
                [this.maxHeap[largest], this.maxHeap[index]];
            index = largest;
        }
    }

    /**
     * Sifts down an element in the min heap
     * @param {number} index - Starting index for sift down operation
     */
    siftDownMin(index) {
        const length = this.minHeap.length;
        while (true) {
            let smallest = index;
            const left = 2 * index + 1;
            const right = 2 * index + 2;

            if (left < length && this.minHeap[left] < this.minHeap[smallest]) {
                smallest = left;
            }
            if (right < length && this.minHeap[right] < this.minHeap[smallest]) {
                smallest = right;
            }

            if (smallest === index) break;

            [this.minHeap[index], this.minHeap[smallest]] = 
                [this.minHeap[smallest], this.minHeap[index]];
            index = smallest;
        }
    }
}

// Example usage:
/*
const medianFinder = new MedianFinder();
medianFinder.addNum(1);    // arr = [1]
medianFinder.addNum(2);    // arr = [1, 2]
console.log(medianFinder.findMedian()); // returns 1.5
medianFinder.addNum(3);    // arr = [1, 2, 3]
console.log(medianFinder.findMedian()); // returns 2.0
*/

// Export the MedianFinder class
module.exports = MedianFinder;
```

This implementation provides an efficient solution for finding the median in a data stream using two heaps. Here are the key features:

1. Uses two heaps to maintain the median:
   - A max heap for the lower half of numbers
   - A min heap for the upper half of numbers

2. The `addNum` method efficiently adds new numbers while maintaining the heap properties.

3. The `findMedian` method returns the median in O(1) time:
   - If both heaps have equal size, returns average of top elements
   - Otherwise, returns top element of max heap

4. Includes helper methods for heap operations:
   - `maxHeapify` and `minHeapify` for maintaining heap properties
   - `balanceHeaps` for ensuring proper distribution of elements
   - `siftDownMax` and `siftDownMin` for heap maintenance

5. Handles edge cases appropriately and maintains proper heap balance.

6. Includes clear comments and follows JavaScript best practices.

The time complexity is O(log n) for insertions and O(1) for finding the median, where n is the number of elements in the data structure.